### DISCUSSION TITLE
[Feature Request] Add String Functions to Grafana

### DISCUSSION THREAD

**Discussion by anybodysguest** [grafana/grafana#37764] [Dashboards, Panels & Visualizations]
There are many strengths of Grafana, but one strength that stands above the crowd is "templating".

However, as powerful as templating is, they are inflexible in the sense that the values of template variables cannot be easily be tweaked for certain situations.  Advanced Variable Format options is a hint of potential better things to come, but in it's present for it's not that useful.

I propose that addition functions be added the the Advance Variable Format Options(AVFO) to include string function that would allow one to leverage the power of templated variables.

The the functions would be "chained" together, interpreted from the left to right allowing for multiple functions to be used together in one command.

Variable Name: testvar
Chained function syntax: var:function1.function2....functionN
"." between functions would essentially pipe the output from one function to the next.

Chaining example 1: ${testvar:substring(1,4).lcase} or ${testvar:lcase.substring(1,4)}
Interpretation:  
 1. Substring function returns "Test" and passes this on to the next function.
 2. Lcase takes the output from substring and lower cases the value to "test"
 3. Interpolation: "test"

**Proposed functions:**
For the examples below we have a template constant variable with a value of "development", the variable name will be "environment"

**len** - returns the length of a variable value
  Example: "The length of the variable is: ${environment:len}"
  Interpolation Result: "Then length of the variable is: 11"

**substring** - returns a portion of a variable value
 Example: "Your application is currently running in the '${environment:substring(1,3)}' environment"
 Interpolation Result: "Your application is currently running in the  'dev' environment"

**ucase/lcase** - Upper/Lower cases the value of a variable's value
 Example: "Current environment is: ${environment:ucase}"
 Interpolation Result: "Current environment is: DEVELOPMENT"

**append/prepend** - Appends a value to a variable value or to a function's output
 Example1: "Current environment is: ${environment:ucase.append('-'${region.ucase})}"
 assumption: region variable = "us-west-2"
 Interpolation Result: "Current environment is: DEVELOPMENT-US-WEST-2"
OR
 Example2: "Current environment is: ${environment:ucase.append('-SERVER')}"
 Interpolation Result: "Current environment is: DEVELOPMENT-SERVER"
OR
 Example3: "Current environment is: ${environment:prepend('CURRENT-ENV-')}"
 Interpolation Result: "Current environment is: CURRENT-ENV-development"

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
